Reading_a_JWST_ASDF_file.ipynb

```diff
@@ -0,0 +1,366 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Outline and Introduction\n",
+    "\n",
+    "- Examining the structure\n",
+    "- Opening an ASDF file\n",
+    "- Summarizing the structure of the file\n",
+    "- Accessing the contents\n",
+    "- Searching for content\n",
+    "\n",
+    "### Basic organization of ASDF files\n",
+    "\n",
+    "Understanding the basic organizational structures of an ASDF file will help in navigating the contents of such files. In brief, the organization consists of an arbitrary nesting of the equivalent of Python lists and dictionaries. As a result, getting to any particular item involves either indexing by number when a list is encountered (YAML parlance is an \"array\") or by a key when a dictionary is encountered (YAML parlance is \"map\"). Furthermore, if any item is preceded by a YAML \"tag\", Python generally knows (if the appropriate software is installed) how to turn it into a corresponding Python object. The most common instance of this case is for ndarray or table data. If a tag for a ndarray is encountered, Python will turn this into a numpy array automatically (though perhaps not until you wish to access the array). But many other kinds of tags and corresponding Python objects will be encountered.\n",
+    "\n",
+    "### Reading a JWST ASDF file\n",
+    "\n",
+    "Unbeknownst to many, while the JWST pipeline normally works with FITS files, it has been designed to work with ASDF files as well (though there is currently a small change needed to enable this). Turning JWST data products into ASDF files is easy, and our first example will be with such a file. First some preliminaries such as doing the needed imports and references to the data files used for this tutorial (they must be downloaded, but this notebook handles that)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import asdf\n",
+    "from astropy.utils.data import download_file\n",
+    "import matplotlib\n",
+    "matplotlib.rcParams['figure.figsize'] = (10, 10)\n",
+    "from matplotlib import pyplot as plt\n",
+    "file1_url_1 = 'https://data.science.stsci.edu/redirect/ASDF/asdf_tut_1.asdf'\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# This is a large file and may take a while to download\n",
+    "filepath = download_file(file1_url_1)\n",
+    "# Wait until download completes (when 'In [*]' changes the  '* to a number')\n",
+    "# Ignore warnings"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": []
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Quick look at contents"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "af.info()"
```

Comment:
One of the cells must have gotten lost, this fails with:

```
NameError: name 'af' is not defined
```

Reply:
I hunted it down and dragged it back home (along with a stern talking to).